includes/panopto_derivatives.inc

```diff
@@ -5,37 +5,356 @@
+          'type' => 'watchdog',
+          'severity' => WATCHDOG_ERROR,
+        );
+        $failureMessage['messages'][] = $messages;
+      }
+    } else {
+      $messages = array(
+        'message' => t("For object @pid : Expected one of " .implode(',', array_keys($expect))."; got a $response_code"),
+        'message_sub' => array('@pid' => $object->id), 
+        'type' => 'watchdog',
+        'severity' => WATCHDOG_ERROR,
+      );
+      $failureMessage['messages'][] = $messages;
+    }
+  }
+  return;
+}
+
+/*
+ * Uploads file to Panopto using AWS S3 Uploader
+ * @param $upload_target - url of upload target
+ * @param $file_path - path of object
+ */
+function islandora_video_streaming_panopto_upload_file($upload_target, $file_path) {
+  $element = explode("/", $upload_target);
+  $prefix = array_pop($element);
+  $service_endpoint = implode("/", $element);
+  $bucket = array_pop($element);
+  $file_name = basename($file_path);
+  $object_key = $prefix."/".$file_name;
+  
+  $s3Params = array(
+      'endpoint' => $service_endpoint,
+      'region'  => 'us-east-1',
+      'version' => '2006-03-01',
+      'credentials' => [
+        'key'    => 'dummy',
+        'secret' => 'dummy'
+      ]
+  );
+  // Create an S3Client
+  $s3Client = new S3Client($s3Params);
+
+  $source = fopen($file_path, 'rb');
+  $uploader = new ObjectUploader(
+    $s3Client,
+    $bucket,
+    $object_key,
+    $source
+  );
+
+  do {
+    try {
+      $result = $uploader->upload();
+    } catch (Exception $ex) {
+      $messages = array(
+        'message' => t('S3Client and Uploader error: @exception'),
+        'message_sub' => array('@exception' => $ex->getMessage()),
+        'type' => 'watchdog',
+        'severity' => WATCHDOG_ERROR,
+      );
+      $failureMessage['messages'][] = $messages;
+      watchdog_exception('islandora_video_streaming_panopto', $ex);
+    }
+  } while (!isset($result));
+}
+
+/*
+ * Creates manifest file, in temporary storage, needed for Panopto Video Upload
+ * @param $file_path
+ * @param $manifest_file_name
+ */
+function islandora_video_streaming_panopto_create_manifest_for_video($file_path = null, $manifest_file_name=null) {
+  $file_name = basename($file_path);
+
+  $template = file_get_contents('https://raw.githubusercontent.com/Panopto/upload-python-sample/master/simplest/upload_manifest_template.xml');
+  $template = str_replace("{Title}", $file_name, $template);
+  $template = str_replace("{Description}", 'This is a video session with the uploaded video file '.$file_name, $template);
+  $template = str_replace("{Filename}", $file_name, $template);
+  $template = str_replace("{Date}", date('c'), $template);
+  
+  $finished = file_put_contents(MANIFEST_FILE_NAME, $template);
+}
+
+/*
+ * Finishes upload
+ * @param $baseUrl
+ * @param $session_upload
+ * @param $token
+ */
+function islandora_video_streaming_panopto_finish_upload($baseUrl, $session_upload, $token) {
+ $upload_id = $session_upload['ID'];
```

Comment:
This function is also extremely light on error checking, but you can probably get away with it, if `islandora_video_streaming_panopto_request_curl` handles error checking well.  You may get some undefined or casting   warnings if `$baseUrl`, `$sessionUpload` or `$token` are not what you expect them to be.

Reply:
I was under the assumption if the program gets to this part of the code, it should have a valid `$sessionUpload', '$baseUrl', and '$token`.

Reply:
This may have a valid `$baseUrl`, `$session_upload`, and `$token` for calls from *your* existent code, but you have to account for everyone else's possible future code, including your own.